File: sdk/nodejs/tests/autoscaler.test.ts
/**
 * Mock Tests for CAST AI Autoscaler Resource (TypeScript)
 *
 * These tests use Pulumi's built-in mocking to test autoscaler configuration
 * without making actual API calls to CAST AI.
 *
 * Test scenarios are based on the Terraform provider examples at:
 * terraform-provider-castai/examples/eks/eks_cluster_existing/castai.tf
 *
 * Run with: npm test
 */

import * as pulumi from "@pulumi/pulumi";
import * as castai from "@castai/pulumi";
import { promisify, promisifyAll } from "./test-utils";

/**
 * Mock implementation for CAST AI Autoscaler resources.
 */
class CastAIMocks implements pulumi.runtime.Mocks {
    newResource(args: pulumi.runtime.MockResourceArgs): { id: string; state: any } {
        const outputs = { ...args.inputs };

        if (args.type === "castai:autoscaling:Autoscaler") {
            return {
                id: `${args.name}-autoscaler-id-${this.hash(args.name)}`,
                state: {
                    ...outputs,
                    id: `${args.name}-autoscaler-id-${this.hash(args.name)}`,
                    autoscalerPolicies: outputs.autoscalerPoliciesJson || "{}",
                },
            };
        }

        return {
            id: `${args.name}-id`,
            state: outputs,
        };
    }

    call(args: pulumi.runtime.MockCallArgs): Record<string, any> {
        return {};
    }

    private hash(str: string): number {
        let hash = 0;
        for (let i = 0; i < str.length; i++) {
            const char = str.charCodeAt(i);
            hash = (hash << 5) - hash + char;
            hash = hash & hash;
        }
        return Math.abs(hash) % 1000;
    }
}

// Set up mocks
pulumi.runtime.setMocks(new CastAIMocks());

describe("Autoscaler Configuration Tests", () => {
    it("should create autoscaler with basic disabled configuration", async () => {
        // Based on Terraform example: disabled autoscaler
        const autoscalerPolicies = {
            enabled: false,
            isScopedMode: false,
            nodeTemplatesPartialMatchingEnabled: false,
            unschedulablePods: {
                enabled: false,
            },
            nodeDownscaler: {
                enabled: false,
                emptyNodes: {
                    enabled: false,
                },
                evictor: {
                    enabled: false,
                    aggressiveMode: false,
                    cycleInterval: "60s",
                    nodeGracePeriodMinutes: 10,
                    scopedMode: false,
                },
            },
            clusterLimits: {
                enabled: false,
                cpu: {
                    maxCores: 20,
                    minCores: 1,
                },
                spotBackups: {
                    enabled: false,
                    spotBackupRestoreRateSeconds: 1800,
                },
            },
        };

        const autoscaler = new castai.Autoscaler("test-autoscaler", {
            clusterId: "test-cluster-id-123",
            autoscalerPoliciesJson: JSON.stringify(autoscalerPolicies),
        });

        const [clusterId, policiesJson] = await promisifyAll(
            autoscaler.clusterId,
            autoscaler.autoscalerPoliciesJson
        );

        expect(clusterId).toBe("test-cluster-id-123");
        expect(policiesJson).toBeDefined();

        const policies = JSON.parse(policiesJson!);
        expect(policies.enabled).toBe(false);
        expect(policies.clusterLimits.cpu.maxCores).toBe(20);
    });

    it("should create autoscaler with enabled features", async () => {
        const autoscalerPolicies = {
            enabled: true,
            isScopedMode: false,
            nodeTemplatesPartialMatchingEnabled: false,
            unschedulablePods: {
                enabled: true,
            },
            nodeDownscaler: {
                enabled: true,
                emptyNodes: {
                    enabled: true,
                },
                evictor: {
                    enabled: true,
                    aggressiveMode: false,
                    cycleInterval: "5m10s",
                    nodeGracePeriodMinutes: 10,
                    scopedMode: false,
                },
            },
            clusterLimits: {
                enabled: true,
                cpu: {
                    maxCores: 100,
                    minCores: 2,
                },
                spotBackups: {
                    enabled: true,
                    spotBackupRestoreRateSeconds: 1800,
                },
            },
        };

        const autoscaler = new castai.Autoscaler("test-autoscaler-enabled", {
            clusterId: "enabled-cluster-123",
            autoscalerPoliciesJson: JSON.stringify(autoscalerPolicies),
        });

        const [clusterId, policiesJson] = await promisifyAll(
            autoscaler.clusterId,
            autoscaler.autoscalerPoliciesJson
        );

        expect(clusterId).toBe("enabled-cluster-123");

        const policies = JSON.parse(policiesJson!);
        expect(policies.enabled).toBe(true);
        expect(policies.unschedulablePods.enabled).toBe(true);
        expect(policies.nodeDownscaler.enabled).toBe(true);
        expect(policies.nodeDownscaler.evictor.enabled).toBe(true);
        expect(policies.clusterLimits.enabled).toBe(true);
        expect(policies.clusterLimits.cpu.maxCores).toBe(100);
    });

    it("should create autoscaler with aggressive evictor mode", async () => {
        const autoscalerPolicies = {
            enabled: true,
            nodeDownscaler: {
                enabled: true,
                evictor: {
                    enabled: true,
                    aggressiveMode: true, // Aggressive mode enabled
                    cycleInterval: "30s", // Faster cycle
                    nodeGracePeriodMinutes: 5, // Shorter grace period
                    scopedMode: false,
                },
            },
        };

        const autoscaler = new castai.Autoscaler("test-aggressive-evictor", {
            clusterId: "aggressive-cluster-456",
            autoscalerPoliciesJson: JSON.stringify(autoscalerPolicies),
        });

        const [clusterId, policiesJson] = await promisifyAll(
            autoscaler.clusterId,
            autoscaler.autoscalerPoliciesJson
        );

        const policies = JSON.parse(policiesJson!);
        const evictor = policies.nodeDownscaler.evictor;
        expect(evictor.enabled).toBe(true);
        expect(evictor.aggressiveMode).toBe(true);
        expect(evictor.cycleInterval).toBe("30s");
        expect(evictor.nodeGracePeriodMinutes).toBe(5);
    });

    it("should create autoscaler with various cluster limits", async () => {
        const autoscalerPolicies = {
            enabled: true,
            clusterLimits: {
                enabled: true,
                cpu: {
                    maxCores: 50,
                    minCores: 5,
                },
                spotBackups: {
                    enabled: true,
                    spotBackupRestoreRateSeconds: 3600, // 1 hour
                },
            },
        };

        const autoscaler = new castai.Autoscaler("test-cluster-limits", {
            clusterId: "limited-cluster-789",
            autoscalerPoliciesJson: JSON.stringify(autoscalerPolicies),
        });

        const [clusterId, policiesJson] = await promisifyAll(
            autoscaler.clusterId,
            autoscaler.autoscalerPoliciesJson
        );

        const policies = JSON.parse(policiesJson!);
        const limits = policies.clusterLimits;
        expect(limits.enabled).toBe(true);
        expect(limits.cpu.maxCores).toBe(50);
        expect(limits.cpu.minCores).toBe(5);
        expect(limits.spotBackups.spotBackupRestoreRateSeconds).toBe(3600);
    });

    it("should create autoscaler with scoped mode enabled", async () => {
        const autoscalerPolicies = {
            enabled: true,
            isScopedMode: true, // Scoped mode - only manages specific namespaces
            nodeTemplatesPartialMatchingEnabled: true,
            unschedulablePods: {
                enabled: true,
            },
            nodeDownscaler: {
                enabled: true,
                evictor: {
                    enabled: true,
                    scopedMode: true, // Evictor also in scoped mode
                },
            },
        };

        const autoscaler = new castai.Autoscaler("test-scoped-mode", {
            clusterId: "scoped-cluster-321",
            autoscalerPoliciesJson: JSON.stringify(autoscalerPolicies),
        });

        const [clusterId, policiesJson] = await promisifyAll(
            autoscaler.clusterId,
            autoscaler.autoscalerPoliciesJson
        );

        const policies = JSON.parse(policiesJson!);
        expect(policies.isScopedMode).toBe(true);
        expect(policies.nodeTemplatesPartialMatchingEnabled).toBe(true);
        expect(policies.nodeDownscaler.evictor.scopedMode).toBe(true);
    });

    it("should create autoscaler with minimal configuration", async () => {
        // Minimal configuration - just enable basic autoscaling
        const autoscalerPolicies = {
            enabled: true,
        };

        const autoscaler = new castai.Autoscaler("test-minimal", {
            clusterId: "minimal-cluster-999",
            autoscalerPoliciesJson: JSON.stringify(autoscalerPolicies),
        });

        const [clusterId, policiesJson] = await promisifyAll(
            autoscaler.clusterId,
            autoscaler.autoscalerPoliciesJson
        );

        expect(clusterId).toBe("minimal-cluster-999");

        const policies = JSON.parse(policiesJson!);
        expect(policies.enabled).toBe(true);
    });

    it("should create autoscaler with production-ready settings", async () => {
        // Production settings: enabled but conservative
        const autoscalerPolicies = {
            enabled: true,
            isScopedMode: false,
            nodeTemplatesPartialMatchingEnabled: false,
            unschedulablePods: {
                enabled: true,
            },
            nodeDownscaler: {
                enabled: true,
                emptyNodes: {
                    enabled: true,
                },
                evictor: {
                    enabled: true,
                    aggressiveMode: false, // Conservative for production
                    cycleInterval: "10m", // Longer cycle for stability
                    nodeGracePeriodMinutes: 15, // Longer grace period
                    scopedMode: false,
                },
            },
            clusterLimits: {
                enabled: true,
                cpu: {
                    maxCores: 200, // Higher limit for production
                    minCores: 10, // Ensure baseline capacity
                },
                spotBackups: {
                    enabled: true,
                    spotBackupRestoreRateSeconds: 1800,
                },
            },
        };

        const autoscaler = new castai.Autoscaler("prod-autoscaler", {
            clusterId: "prod-cluster-001",
            autoscalerPoliciesJson: JSON.stringify(autoscalerPolicies),
        });

        const [clusterId, policiesJson] = await promisifyAll(
            autoscaler.clusterId,
            autoscaler.autoscalerPoliciesJson
        );

        expect(clusterId).toBe("prod-cluster-001");

        const policies = JSON.parse(policiesJson!);
        expect(policies.enabled).toBe(true);
        expect(policies.nodeDownscaler.evictor.aggressiveMode).toBe(false);
        expect(policies.nodeDownscaler.evictor.nodeGracePeriodMinutes).toBe(15);
        expect(policies.clusterLimits.cpu.minCores).toBe(10);
    });
});

File: sdk/go/tests/autoscaler_test.go
package tests

import (
	"encoding/json"
	"fmt"
	"testing"

	"github.com/castai/pulumi-castai/sdk/go/castai"
	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
	"github.com/stretchr/testify/assert"
)

// AutoscalerMocks implements pulumi.MockResourceMonitor for Autoscaler resources
type AutoscalerMocks struct {
	pulumi.MockResourceMonitor
}

// NewResource mocks resource creation for autoscaler
func (m *AutoscalerMocks) NewResource(args pulumi.MockResourceArgs) (string, resource.PropertyMap, error) {
	outputs := args.Inputs.Copy()

	switch args.TypeToken {
	case "castai:autoscaling:Autoscaler":
		outputs["id"] = resource.NewStringProperty(fmt.Sprintf("%s-autoscaler-id-%d", args.Name, hashString(args.Name)))
		if policiesJson, ok := outputs["autoscalerPoliciesJson"]; ok {
			outputs["autoscalerPolicies"] = policiesJson
		} else {
			outputs["autoscalerPolicies"] = resource.NewStringProperty("{}")
		}
	default:
		outputs["id"] = resource.NewStringProperty(fmt.Sprintf("%s-id", args.Name))
	}

	return args.Name + "-id", outputs, nil
}

// Call mocks function/data source calls
func (m *AutoscalerMocks) Call(args pulumi.MockCallArgs) (resource.PropertyMap, error) {
	return resource.PropertyMap{}, nil
}

func TestAutoscalerBasicConfiguration(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		// Based on Terraform example: disabled autoscaler
		autoscalerPolicies := map[string]interface{}{
			"enabled":                             false,
			"isScopedMode":                        false,
			"nodeTemplatesPartialMatchingEnabled": false,
			"unschedulablePods": map[string]interface{}{
				"enabled": false,
			},
			"nodeDownscaler": map[string]interface{}{
				"enabled": false,
				"emptyNodes": map[string]interface{}{
					"enabled": false,
				},
				"evictor": map[string]interface{}{
					"enabled":                 false,
					"aggressiveMode":          false,
					"cycleInterval":           "60s",
					"nodeGracePeriodMinutes":  10,
					"scopedMode":              false,
				},
			},
			"clusterLimits": map[string]interface{}{
				"enabled": false,
				"cpu": map[string]interface{}{
					"maxCores": 20,
					"minCores": 1,
				},
				"spotBackups": map[string]interface{}{
					"enabled":                        false,
					"spotBackupRestoreRateSeconds":   1800,
				},
			},
		}

		policiesJSON, err := json.Marshal(autoscalerPolicies)
		assert.NoError(t, err)

		autoscaler, err := castai.NewAutoscaler(ctx, "test-autoscaler", &castai.AutoscalerArgs{
			ClusterId:               pulumi.String("test-cluster-id-123"),
			AutoscalerPoliciesJson:  pulumi.String(string(policiesJSON)),
		})
		assert.NoError(t, err)
		assert.NotNil(t, autoscaler)

		return nil
	}, pulumi.WithMocks("project", "stack", &AutoscalerMocks{}))

	assert.NoError(t, err)
}

func TestAutoscalerEnabledConfiguration(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		autoscalerPolicies := map[string]interface{}{
			"enabled": true,
			"isScopedMode": false,
			"nodeTemplatesPartialMatchingEnabled": false,
			"unschedulablePods": map[string]interface{}{
				"enabled": true,
			},
			"nodeDownscaler": map[string]interface{}{
				"enabled": true,
				"emptyNodes": map[string]interface{}{
					"enabled": true,
				},
				"evictor": map[string]interface{}{
					"enabled":                true,
					"aggressiveMode":         false,
					"cycleInterval":          "5m10s",
					"nodeGracePeriodMinutes": 10,
					"scopedMode":             false,
				},
			},
			"clusterLimits": map[string]interface{}{
				"enabled": true,
				"cpu": map[string]interface{}{
					"maxCores": 100,
					"minCores": 2,
				},
				"spotBackups": map[string]interface{}{
					"enabled":                      true,
					"spotBackupRestoreRateSeconds": 1800,
				},
			},
		}

		policiesJSON, err := json.Marshal(autoscalerPolicies)
		assert.NoError(t, err)

		autoscaler, err := castai.NewAutoscaler(ctx, "test-autoscaler-enabled", &castai.AutoscalerArgs{
			ClusterId:              pulumi.String("enabled-cluster-123"),
			AutoscalerPoliciesJson: pulumi.String(string(policiesJSON)),
		})
		assert.NoError(t, err)
		assert.NotNil(t, autoscaler)

		return nil
	}, pulumi.WithMocks("project", "stack", &AutoscalerMocks{}))

	assert.NoError(t, err)
}

func TestAutoscalerAggressiveEvictor(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		autoscalerPolicies := map[string]interface{}{
			"enabled": true,
			"nodeDownscaler": map[string]interface{}{
				"enabled": true,
				"evictor": map[string]interface{}{
					"enabled":                true,
					"aggressiveMode":         true, // Aggressive mode enabled
					"cycleInterval":          "30s", // Faster cycle
					"nodeGracePeriodMinutes": 5, // Shorter grace period
					"scopedMode":             false,
				},
			},
		}

		policiesJSON, err := json.Marshal(autoscalerPolicies)
		assert.NoError(t, err)

		autoscaler, err := castai.NewAutoscaler(ctx, "test-aggressive-evictor", &castai.AutoscalerArgs{
			ClusterId:              pulumi.String("aggressive-cluster-456"),
			AutoscalerPoliciesJson: pulumi.String(string(policiesJSON)),
		})
		assert.NoError(t, err)
		assert.NotNil(t, autoscaler)

		return nil
	}, pulumi.WithMocks("project", "stack", &AutoscalerMocks{}))

	assert.NoError(t, err)
}

func TestAutoscalerClusterLimits(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		autoscalerPolicies := map[string]interface{}{
			"enabled": true,
			"clusterLimits": map[string]interface{}{
				"enabled": true,
				"cpu": map[string]interface{}{
					"maxCores": 50,
					"minCores": 5,
				},
				"spotBackups": map[string]interface{}{
					"enabled":                      true,
					"spotBackupRestoreRateSeconds": 3600, // 1 hour
				},
			},
		}

		policiesJSON, err := json.Marshal(autoscalerPolicies)
		assert.NoError(t, err)

		autoscaler, err := castai.NewAutoscaler(ctx, "test-cluster-limits", &castai.AutoscalerArgs{
			ClusterId:              pulumi.String("limited-cluster-789"),
			AutoscalerPoliciesJson: pulumi.String(string(policiesJSON)),
		})
		assert.NoError(t, err)
		assert.NotNil(t, autoscaler)

		return nil
	}, pulumi.WithMocks("project", "stack", &AutoscalerMocks{}))

	assert.NoError(t, err)
}

func TestAutoscalerScopedMode(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		autoscalerPolicies := map[string]interface{}{
			"enabled":     true,
			"isScopedMode": true, // Scoped mode - only manages specific namespaces
			"nodeTemplatesPartialMatchingEnabled": true,
			"unschedulablePods": map[string]interface{}{
				"enabled": true,
			},
			"nodeDownscaler": map[string]interface{}{
				"enabled": true,
				"evictor": map[string]interface{}{
					"enabled":    true,
					"scopedMode": true, // Evictor also in scoped mode
				},
			},
		}

		policiesJSON, err := json.Marshal(autoscalerPolicies)
		assert.NoError(t, err)

		autoscaler, err := castai.NewAutoscaler(ctx, "test-scoped-mode", &castai.AutoscalerArgs{
			ClusterId:              pulumi.String("scoped-cluster-321"),
			AutoscalerPoliciesJson: pulumi.String(string(policiesJSON)),
		})
		assert.NoError(t, err)
		assert.NotNil(t, autoscaler)

		return nil
	}, pulumi.WithMocks("project", "stack", &AutoscalerMocks{}))

	assert.NoError(t, err)
}

func TestAutoscalerMinimalConfiguration(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		// Minimal configuration - just enable basic autoscaling
		autoscalerPolicies := map[string]interface{}{
			"enabled": true,
		}

		policiesJSON, err := json.Marshal(autoscalerPolicies)
		assert.NoError(t, err)

		autoscaler, err := castai.NewAutoscaler(ctx, "test-minimal", &castai.AutoscalerArgs{
			ClusterId:              pulumi.String("minimal-cluster-999"),
			AutoscalerPoliciesJson: pulumi.String(string(policiesJSON)),
		})
		assert.NoError(t, err)
		assert.NotNil(t, autoscaler)

		return nil
	}, pulumi.WithMocks("project", "stack", &AutoscalerMocks{}))

	assert.NoError(t, err)
}

func TestAutoscalerProductionSettings(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		// Production settings: enabled but conservative
		autoscalerPolicies := map[string]interface{}{
			"enabled":                             true,
			"isScopedMode":                        false,
			"nodeTemplatesPartialMatchingEnabled": false,
			"unschedulablePods": map[string]interface{}{
				"enabled": true,
			},
			"nodeDownscaler": map[string]interface{}{
				"enabled": true,
				"emptyNodes": map[string]interface{}{
					"enabled": true,
				},
				"evictor": map[string]interface{}{
					"enabled":                true,
					"aggressiveMode":         false, // Conservative for production
					"cycleInterval":          "10m", // Longer cycle for stability
					"nodeGracePeriodMinutes": 15,   // Longer grace period
					"scopedMode":             false,
				},
			},
			"clusterLimits": map[string]interface{}{
				"enabled": true,
				"cpu": map[string]interface{}{
					"maxCores": 200, // Higher limit for production
					"minCores": 10,  // Ensure baseline capacity
				},
				"spotBackups": map[string]interface{}{
					"enabled":                      true,
					"spotBackupRestoreRateSeconds": 1800,
				},
			},
		}

		policiesJSON, err := json.Marshal(autoscalerPolicies)
		assert.NoError(t, err)

		autoscaler, err := castai.NewAutoscaler(ctx, "prod-autoscaler", &castai.AutoscalerArgs{
			ClusterId:              pulumi.String("prod-cluster-001"),
			AutoscalerPoliciesJson: pulumi.String(string(policiesJSON)),
		})
		assert.NoError(t, err)
		assert.NotNil(t, autoscaler)

		return nil
	}, pulumi.WithMocks("project", "stack", &AutoscalerMocks{}))

	assert.NoError(t, err)
}
